protocol/nats_jetstream/v3/receiver.go

```diff
@@ -0,0 +1,219 @@
+/*
+ Copyright 2021 The CloudEvents Authors
+ SPDX-License-Identifier: Apache-2.0
+*/
+
+package nats_jetstream
+
+import (
+	"context"
+	"io"
+	"sync"
+
+	"github.com/nats-io/nats.go"
+	"github.com/nats-io/nats.go/jetstream"
+
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/protocol"
+)
+
+type msgErr struct {
+	msg binding.Message
+}
+
+type receiver struct {
+	incoming chan msgErr
+}
+
+// newReceiver creates a new protocol.Receiver responsible for receiving messages.
+func newReceiver() *receiver {
+	return &receiver{
+		incoming: make(chan msgErr),
+	}
+}
+
+// MsgHandler implements nats.MsgHandler and publishes messages onto our internal incoming channel to be delivered
+// via r.Receive(ctx)
+func (r *receiver) MsgHandler(msg jetstream.Msg) {
+	r.incoming <- msgErr{msg: NewMessage(msg)}
+}
+
+// Receive implements Receiver.Receive.
+func (r *receiver) Receive(ctx context.Context) (binding.Message, error) {
+	select {
+	case msgErr, ok := <-r.incoming:
+		if !ok {
+			return nil, io.EOF
+		}
+		return msgErr.msg, nil
+	case <-ctx.Done():
+		return nil, io.EOF
+	}
+}
+
+type consumer struct {
+	receiver
```

Comment:
It seems that there are too many layers of nesting in the implementation of the `protocol.Receiver`. (Protocol -> consumer -> receiver). This deep nesting appears to increase the complexity of the code maintenance